controller/cache/scan.go

```
@@ -368,12 +368,15 @@ func scanObject(id string) {
 }
 
 func (m CacheMethod) ScanWorkload(id string, acc *access.AccessControl) error {
-	if cache := getWorkloadCache(id); cache == nil {
+	var cache *workloadCache
+	if cache = getWorkloadCache(id); cache == nil {
 		return common.ErrObjectNotFound
 	} else if !acc.Authorize(&share.CLUSWorkloadScanDummy{Domain: cache.workload.Domain}, nil) {
 		return common.ErrObjectAccessDenied
 	}
 
+	// Ensure the workload object is in scanMap
+	scanWorkloadAdd(id, cache)
```

If the agent is gone on the host of the workload, line 379 could add the workload to scanMap but there is still no agent for scanning the workload & the caller of ScanWorkload() still gets a nil error. 
Why not just return an error to the caller when the agent is unavailable ?

---

I think you are right about this.
However, if the agent is gone, the scan object should encounter the error.

This line just to ensure the sync of scanMap and wlcacheMap.

What do you think?

---

I think we should not call scanWorkloadAdd(id, cache) unconditionally.
Could we check the agent of the workload(by `cache`) is in agentCacheMap or not before calling scanWorkloadAdd(id, cache)?
(If not found, return error instead of keep calling scanObject(id) so that caller knows the `scan workload` fails with reason.)

---

when agent is down there is a 10 minute wait before agent is really removed from cache, the logic is complex, pls check this function deleteAgentFromCache for reference